azuread/configuration.go

```diff
@@ -136,18 +149,81 @@ func (p *azureFedAuthConfig) validateParameters(params map[string]string) error
+		// Split the clientID@tenantID format
+		p.clientID, p.tenantID = splitTenantAndClientID(params["user id"])
+		if p.clientID == "" {
+			return errors.New("Must provide 'client id[@tenant id]' as username parameter when using ActiveDirectoryClientAssertion authentication")
+		}
+		p.clientAssertion = params["clientassertion"]
+		if p.clientAssertion == "" {
+			return errors.New("Must provide 'clientassertion' parameter when using ActiveDirectoryClientAssertion authentication")
+		}
+	case strings.EqualFold(fedAuthWorkflow, ActiveDirectoryOnBehalfOf):
+		p.adalWorkflow = mssql.FedAuthADALWorkflowPassword
+		// Split the clientID@tenantID format
+		p.clientID, p.tenantID = splitTenantAndClientID(params["user id"])
+		if p.clientID == "" {
+			return errors.New("Must provide 'client id[@tenant id]' as username parameter when using ActiveDirectoryOnBehalfOf authentication")
+		}
+		p.userAssertion = params["userassertion"]
+		if p.userAssertion == "" {
+			return errors.New("Must provide 'userassertion' parameter when using ActiveDirectoryOnBehalfOf authentication")
+		}
+		// On-behalf-of can use client secret, certificate, or client assertion
+		p.clientSecret = params["password"]
+		p.certificatePath = params["clientcertpath"]
+		p.clientAssertion = params["clientassertion"]
+		if p.clientSecret == "" && p.certificatePath == "" && p.clientAssertion == "" {
+			return errors.New("Must provide one of 'password', 'clientcertpath', or 'clientassertion' parameter when using ActiveDirectoryOnBehalfOf authentication")
 		}
-	default:
-		return fmt.Errorf("Invalid federated authentication type '%s': expected one of %+v",
-			fedAuthWorkflow,
-			[]string{ActiveDirectoryApplication, ActiveDirectoryServicePrincipal, ActiveDirectoryDefault, ActiveDirectoryIntegrated, ActiveDirectoryInteractive, ActiveDirectoryManagedIdentity, ActiveDirectoryMSI, ActiveDirectoryPassword, ActiveDirectoryAzCli, ActiveDirectoryDeviceCode})
+	default:
+		return fmt.Errorf("Invalid federated authentication type '%s': expected one of %+v",
+			fedAuthWorkflow,
+			[]string{ActiveDirectoryApplication, ActiveDirectoryServicePrincipal, ActiveDirectoryDefault, ActiveDirectoryIntegrated, ActiveDirectoryInteractive, ActiveDirectoryManagedIdentity, ActiveDirectoryMSI, ActiveDirectoryPassword, ActiveDirectoryAzCli, ActiveDirectoryDeviceCode, ActiveDirectoryAzureDeveloperCli, ActiveDirectoryAzurePipelines, ActiveDirectoryEnvironment, ActiveDirectoryWorkloadIdentity, ActiveDirectoryClientAssertion, ActiveDirectoryOnBehalfOf})
 	}
 	p.fedAuthWorkflow = fedAuthWorkflow
-	return nil
-}
-
+	return nil
+}
+
+// isPasswordWorkflowAuth checks if the federated auth workflow uses the password workflow
+func isPasswordWorkflowAuth(workflow string) bool {
+	passwordWorkflows := []string{
+		ActiveDirectoryDefault,
+		ActiveDirectoryAzCli,
+		ActiveDirectoryDeviceCode,
+		ActiveDirectoryAzureDeveloperCli,
+		ActiveDirectoryEnvironment,
+		ActiveDirectoryWorkloadIdentity,
```

Comment:
>ActiveDirectoryWorkloadIdentity, [](http://example.com/codeflow?start=2&length=32)

`WorkloadIdentityCredentialOptions` can be passed to NewWorkloadIdentityCredential. If the connection string contains a user id like `user@tenant`, we should use that user id as the `ClientID` and `TenantID` parameters in the options.